sweep/sweeper.go

```diff
@@ -670,28 +654,29 @@ func (s *UtxoSweeper) collector(blockEpochs <-chan *chainntnfs.BlockEpoch) {
 				err:        err,
 			}
 
-		// The timer expires and we are going to (re)sweep.
-		case <-ticker.C:
-			log.Debugf("Sweep ticker ticks, attempt sweeping %d "+
-				"inputs", len(inputs))
-
-			// Sweep the remaining pending inputs.
-			s.sweepPendingInputs(inputs)
-
 		// A new block comes in, update the bestHeight.
 		//
 		// TODO(yy): this is where we check our published transactions
 		// and perform RBF if needed. We'd also like to consult our fee
 		// bumper to get an updated fee rate.
 		case epoch, ok := <-blockEpochs:
 			if !ok {
+				// We should stop the sweeper before stopping
+				// the chain service. Otherwise it indicates an
+				// error.
+				log.Error("Block epoch channel closed")
+
 				return
 			}
 
+			// Update the sweeper to the best height.
 			s.currentHeight = epoch.Height
 
-			log.Debugf("New block: height=%v, sha=%v",
-				epoch.Height, epoch.Hash)
+			log.Debugf("Received new block: height=%v, attempt "+
+				"sweeping %d inputs", epoch.Height, len(inputs))
+
+			// Attempt to sweep any pending inputs.
+			s.sweepPendingInputs(inputs)
```

Comment:
If I can crash you every block why couldn't I crash you every 30s? 

I don't see an issue to moving to a timer, though. If you think that makes a material difference, we can probably change that.

Reply:
> If I can crash you every block why couldn't I crash you every 30s?

Maybe, maybe not.  Some DoS attacks cause crashes instantly (e.g., nil dereference, OOB), others overload the victim more slowly and take longer to cause crashes (e.g., OOM, disk-based DoS).

I still think our best defense against DoS is https://github.com/lightningnetwork/lnd/issues/8166.  But even if we have that, we still need to ensure that sweep transactions get broadcast immediately-ish on startup.  If we have to wait 10 minutes, #8166 doesn't help.

Reply:
Hmm don't we have the `Force` Flag in place for time-sensitive sweeps which would let us sweep the inputs during restarts as soon as the resolvers register it with the sweeper ?

Reply:
I think we should fix the DoS and prevent it from happening, rather than trying to build a system assuming there's one - I don't think there's much you can do in that case.

Some users don't use the default 30s, our [docs](https://docs.lightning.engineering/the-lightning-network/payment-channels/understanding-sweeping#docs-internal-guid-fc8a7dce-7fff-508f-ae91-e6f659f04954) recommend 10m, and some may choose hours. And in the given example, if I'm the attacker, I'd crash Alice's node before step 4, rendering her `contractcourt` useless.

> after each restart the victim will wait ~10 minutes (1 block) before attempting to broadcast the sweep transaction instead of 30 seconds.

I think what's missing from our sweeper is the ability to recover from restart - which has been discussed multiple times in these PR series. The sweeper is almost stateless atm, and it should pick up the existing pending inputs during restart, and offer them to the fee bumper immediately.

Reply:
> Designing the system to protect funds under DoS is within reach, so let's do it.

Is it? I am not so sure I agree. The premise you're starting with is that there are DoS vectors that are abundant, and if DoS can take the node down, and there exist deadline based mechanisms in LN, funds will always be at stake. What am I missing?

Reply:
> > Designing the system to protect funds under DoS is within reach, so let's do it.
> 
> Is it? I am not so sure I agree.

#8166 seems well within reach to me.  I think significantly less work is needed than has been done on this set of sweeper PRs already.

>The premise you're starting with is that there are DoS vectors that are abundant, and if DoS can take the node down, and there exist deadline based mechanisms in LN, funds will always be at stake. What am I missing?

Define abundant?  All an attacker needs is 1 DoS vector to cause mayhem with current LND nodes.  Are you confident there are 0 DoS vectors in LND today and forever into the future?

The goal is to keep funds safe even if the node is repeatedly crashed.

> I definitely agree that we should begin sweeping on startup. There is no reason to delay arbitrarily. This seems like an orthogonal issue to whether subsequent sweep operations are block driven or some other heartbeat driven.

Yes, if we set up the contractcourt and sweep on startup before allowing any peer connections, then #8166 has essentially been accomplished.  That's the goal.

Until then, a 30s delay before sweeping on startup seems better than 10 minutes.

Reply:
> All an attacker needs is 1 DoS vector to cause mayhem with current LND nodes. Are you confident there are 0 DoS vectors in LND today and forever into the future?

Absolutely not :).

I'm also not really contesting this. What I'm contesting is what can be done about it. I think #8166 is a generally good idea but it ignores the risks incurred by being disconnected from our peers. I think the assumption that we have to be theoretically crash proof at all times is both unrealistic and a distraction from more pressing issues. I think if you want to make things more crash proof, then decoupling the subsystems is yet another way to handle it. This way crashing the peer interface doesn't take down the whole node, just that particular system.

My main question is that if we're starting from having failure assumptions "someone can crash my node", how broken is reasonable for us to design for? Why not design things in a way where each subsystem can't rely on *any* assumptions from *any other system*? It quickly gets into territory that doesn't make any sense.

I think fault tolerance is a good goal but I also think we need to be careful to not drive ourselves into spending lots of effort on the ever rarer tail events and not accomplish anything while we try to cover edge cases that don't materialize.

I'm open to having my mind changed here, I just generally don't find myself compelled by the argument. One thing that could help me sway is for you to explain why the line in the sand you've happened to draw is unique, and not just an arbitrary point on the continuum towards maximally fault-tolerant software that doesn't consider other dimensions of the problem. Otherwise the slope to me seems rather slippery.

Reply:
> Mallory crashes Alice using some unpatched DoS vector.
> Every time Alice comes back online, Mallory starts DoSing her again. As long as Mallory can keep crashing Alice before the next block confirms, Alice's sweeper will never broadcast any transactions.

So let's say that Alice had a trigger to not accept _any_ connections until all contracts have been resolved. How do we define "resolved" here? Will Alice wait until ultimate _confirmation_ of her sweep? If so, then we also need to be ready for the possibility that there's some crazy fee spike, and confirmation takes much longer than anticipated. If we don't accept _any_ connections during this interval, then it's possible that other active HTLCs we had with other peers, or even other channels with Alice _also_ times out. 

In short, we also need to weigh the opportunity costs of time outs and force closes of _other_ channels if we don't reconnect to any peers until all on chain contracts have been resolved. 

I think what we need to move forward here is a concrete proposal w.r.t what the trigger is to resume reconnections re https://github.com/lightningnetwork/lnd/issues/8166. I think in order to fully isolate we shouldn't even create the listener until we've resolved everything on chain. Of course, kernel level network DoS would still apply, but we only control the application layer in this case.

Reply:
I think we can re-add the timer in addition to the block based model, but in itests, we'd essentially disbale the timer all together (as they were all updated to be block based after this change). 

Before this series of changes, we had the timer as we would at times knowingly broadcast garbage. So we had to keep retrying in the hop that we actually broadcasted something that would confirm, which would then let us retarget to remove the other garbage. With the new approach, we'll no longer broadcast anything that wouldn't be accepted into our node's mempool, so the move to block based (the only time the UTXO state can change) made more sense.

Reply:
I've added a more concrete explanation to https://github.com/lightningnetwork/lnd/issues/8166.  Let's move that discussion over there.

Assuming the new sweeper will be merged before any work on #8166, I think it's worth adding the sweep timer back in for now.  We can remove it once we have #8166.